util/fetch_deps.ts

```diff
@@ -0,0 +1,218 @@
+// execute this script with `npm run fetch-deps`
+import fs from 'fs';
+import path from 'path';
+import crypto from 'crypto';
+
+import _ from 'lodash';
+import tar from 'tar-fs';
+import JSZip from 'jszip';
+import json5 from 'json5';
+import fetch from 'node-fetch';
+import ProxyAgent from 'proxy-agent';
+
+type Meta = {
+  'url': string;
+  'dest': string;
+  'strip': number;
+  'hash': [algorithm: string, hash: string];
+}
+
+const projectRoot = path.resolve(); // path of git repo
+
+const isFile = (p: string): boolean => {
+  if (!fs.existsSync(p))
+    return false;
+  return fs.statSync(p).isFile();
+};
+
+const isDir = (p: string): boolean => {
+  if (!fs.existsSync(p))
+    return false;
+  return fs.statSync(p).isDirectory();
+};
+
+const pad = (x: number) => x.toString().padStart(3, '0');
+
+const endsWith = (s: string, suffix: Iterable<string>): boolean => {
+  for (const ext of suffix) {
+    if (s.endsWith(ext))
+      return true;
+  }
+  return false;
+};
+
+const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));
+
+export const safeRmDir = async (path: string): Promise<void> => {
+  let tries = 30;
+  while (tries > 0) {
+    tries--;
+    try {
+      fs.rmdirSync(path, { recursive: true });
+      break;
+    } catch (e) {
+      if (tries <= 0)
+        throw e;
+      await sleep(300);
+    }
+  }
+};
+
+export const main = async (updateHashes = false): Promise<void> => {
+  const depsPath = path.join(projectRoot, 'util/DEPS.json5');
+
+  const deps = json5.parse<{ [depName: string]: Meta }>(fs.readFileSync(depsPath).toString());
+  let cache: typeof deps = {};
+
+  const cachePath = path.join(projectRoot, 'util/DEPS.cache');
+  const dlPath = path.join(projectRoot, 'util/.deps_dl');
+
+  if (isFile(cachePath))
+    cache = json5.parse<typeof cache>(fs.readFileSync(cachePath).toString());
+
+  const oldCache = new Set(Object.keys(cache));
+  const newCache = new Set(Object.keys(deps));
+
+  const missing = new Set([...newCache].filter((v) => !oldCache.has(v)));
+  const obsolete = new Set([...oldCache].filter((v) => !newCache.has(v)));
+  const outdated: Set<string> = new Set();
+
+  for (const [key, meta] of Object.entries(deps)) {
+    if (!isDir(path.join(projectRoot, meta['dest'])))
+      missing.add(key);
+    else if (oldCache.has(key) && cache[key]?.['hash'][1] !== meta['hash'][1])
+      outdated.add(key);
+  }
+
+  if (isDir(dlPath)) {
+    console.log('Removing left overs...');
+    await safeRmDir(dlPath);
+  }
+
+  fs.mkdirSync(dlPath, { recursive: true });
+
+  const repMap: { [key: string]: string } = {};
+
+  try {
+    const tmp = new Set([...missing, ...outdated]);
+
+    const count = tmp.size;
+    if (count) {
+      console.log('Fetching missing or outdated dependencies...');
+
+      for (const [i, key] of Array.from(tmp.values()).entries()) {
```

Comment:
Could this be simplified to this?

```suggestion
      Array.from(tmp).forEach((key, i) => {
```

(plus the closing paren)

Reply:
No, there are some await async in loops.

Reply:
Ok.  I guess `Array.from(tmp).forEach(async (key, i) => {` is about as long/complicated as what's there.  I think I'm just used to seeing `forEach` when you need the index and not `entries()` destructuring.